man/man5/rlm_sql_log.5
.\"     # DS - begin display
.de DS
.RS
.nf
.sp
..
.\"     # DE - end display
.de DE
.fi
.RE
.sp
..
.TH rlm_sql_log 5 "28 May 2005" "" "FreeRADIUS Module"
.SH NAME
rlm_sql_log \- FreeRADIUS Module
.SH DESCRIPTION
The \fBrlm_sql_log\fP module appends the SQL queries in a log file
which is read later by the scripts/radsqlrelay Perl program.
.PP
The purpose of this module is to de-couple the storage of long-term
accounting data in SQL from "live" information needed by the RADIUS
server as it's running.  If you are not using SQL for simultaneous
login restrictions (i.e. "sql" is not listed in the "session" section
of "radiusd.conf"), then this module allows you to log SQL queries to
a file, and then process them at your leisure.
.PP
The benefit of this approach is that for a busy server, the overhead
of performing SQL qeuries may be significant.  Also, if the SQL
databases are large (as is typical for ones storing months of data),
the INSERTs and UPDATEs may take a relatively long time.  Rather than
slowing down the RADIUS server by having it interact with a database,
you can just log the queries to a file, and then run those queries on
another machine, or at a time when the RADIUS server is typically
lightly loaded.
.PP
If the "sql" module is listed in the "session" section of
"radiusd.conf", then a similar system can still be used.  In that
case, one database would be used to maintain "live" session
information.  That database would be small, fast, and information
would be deleted from it when a user logs out.  A second database
would store long-term accounting information, as described above.
.SH LIMITATIONS
This module only performs the dynamic expansion of the variables found
in the SQL statements. No operation is executed on the database server.
(this would be done later by an external program) That means the module
is useful only with non-"SELECT" statements.
.SH CONFIGURATION
The main configuration items to be aware of are the path of the log
file and the different SQL queries.
.IP "path"
An entry named "path" sets the full path of the file where the SQL
queries are recorded. (this variable is run through dynamic string
expansion, and can include FreeRADIUS variables to create a dynamic
filename)
.IP "Accounting queries"
When a accounting record is processed, the module searches a config
entry keyed by the Acct-Status-Type attribute present in the
packet. For example, the SQL to be run on an accounting start must be
named "Start" in the configuration for the module. Other usual values
for Acct-Status-Type are "Stop", "Alive", "Accounting-On", etc.  See
the VALUEs for Acct-Status-Type in the dictionary.rfc2866 file.
.IP "Post-Auth query"
An entry named "Post-Auth" sets the query to run during the
post-authentication stage.  This query is mainly used to log sessions
where there may not be a later accounting packet.
.PP
.DS
modules {
  ...
.br
  sql_log {
.br
    path = "${radacctdir}/sql-relay"
.br
    acct_table = "radacct"
.br
    postauth_table = "radpostauth"
.br
    sql_user_name = "%{%{User-Name}:-DEFAULT}"
.br

.br
    Start = "INSERT INTO ${acct_table} ..."
.br
    Stop = "UPDATE ${acct_table} SET ..."
.br
    Alive = "UPDATE ${acct_table} SET ..."
.br

.br
    Post-Auth = "INSERT INTO ${postauth_table} ..."
.br
  }
.br
  ...
.br
}

.br
accounting {
  ...
.br
  sql_log
  ...
.br
}
.br

.br
post-auth {
  ...
.br
  sql_log
  ...
.br
}
.DE
.SH SECTIONS
.BR accounting,
.BR post-auth
.SH FILES
.I /etc/raddb/radiusd.conf
.SH SEE ALSO
.BR radsqlrelay (8),
.BR radiusd (8),
.BR radiusd.conf (5)
.SH AUTHOR
Nicolas Baradakis <nicolas.baradakis@cegetel.net>


doc/ChangeLog
FreeRADIUS 2.0.3 ; $Date$ , urgency=medium
	Feature improvements
	* Updated raddb/certs/ca.cnf with extensions to allow ca.der
	  to be imported as a CA on Symbian and Windows Mobile devices.
	  Closes bug #524
	* Enable multiple matches in "hints" via Fall-Through = Yes.
	  Closes bug #477
	* Added preliminary SQLite driver, contibuted by Apple.
	  Untested, with no sample configuration.  This address bug #470.
	* Updated logging sub-system so that log messages from libfreeradius
	  can go to the log file, and not stdout.
	* Added dictionary.rfc5176
	* EAP module now checks for instance name, and uses that for
	  authentication.  This avoids the need to set Auth-Type when
	  there are multiple instances of the EAP module.
	* Added Module-Return-Code attribute, which contains the value
	  returned by the previous module (ok/fail/update/etc.)

	Bug fixes
	* Corrected typos in rlm_dbm.  Closes bugs #521 and #522.
	* Detail file "listen" sections now work much better.
	* Don't allow old "log_*" to over-ride new format.  Closes bug #525
	* Initialize allocated memory in Oracle SQL driver.  This fixes
	  occasional crashes on some systems.  Closes bug #518
	* Call correct function in rlm_protocol_filter.  This enables the
	  module to build.  Closes bug #512.
	* Added deprecated flag to build for rlm_krb5.  This allows it to
	  run on 64-bit systems.  Closes bug #491
	* Corrected error message when parsing invalid configurations
	  so it doesn't crash.  Closes bug #527
	* Fix handling of timeouts in rlm_ldap that affected 64-bit systems.
	* Handle $INCLUDE's in "instantiate" section.  Closes #528.
	* Format updates to "man" pages from Stephen Gran.

FreeRADIUS 2.0.2 ; Date: 2008/02/14 11:13:48  , urgency=medium
	Feature improvements
	* Added notes on how to debug the server in radiusd.conf
	* Moved all "log_*" in radiusd.conf to log{} section.
	  The old configurations are still accepted, though.
	* Added ca.der target in raddb/certs/Makefile.  This is
	  needed for importing CA certs into Windows.
	* Added ability send raw attributes via "Raw-Attribute = 0x0102..."
	  This is available only debug builds.  It can be used
	  to create invalid packets! Use it with care.
	* Permit "unlang" policies inside of Auth-Type{} sub-sections
	  of the authenticate{} section.  This makes some policies easier
	  to implement.
	* "listen" sections can now have "type = proxy".  This lets you
	  control which IP is used for sending proxied requests.
	* Added note on SSL performance to raddb/certs/README

	Bug fixes
	* Fixed reading of "detail" files.
	* Allow inner EAP tunneled sessions to be proxied.
	* Corrected MySQL schemas
	* syslog now works in log{} section.
	* Corrected typo in raddb/certs/client.cnf
	* Updated raddb/sites-available/proxy-inner-tunnel to
	  permit authentication to work.
	* Ignore zero-length attributes in received packets.
	* Correct memcpy when dealing with unknown attributes.
	* Corrected debugging messages in attr_rewrite.
	* Corrected generation of State attribute in EAP.  This
	  fixes the "failed to remember handler" issues.
	* Fall back to DEFAULT realm if no realm was found.
	  Based on a patch from Vincent Magnin.
	* Updated example raddb/sites-available/proxy-inner-tunnel
	* Corrected behavior of attr_filter to match documentation.
	  This is NOT backwards compatible with previous versions!
	  See "man rlm_attr_filter" for details.

FreeRADIUS 2.0.1 ; Date: 2008/01/22 13:29:37  , urgency=low
	Feature improvements
	* "unlang" has been expanded to do less run-time expansion,
	  and to have better handling of typed data.  See "man unlang"
	  for documentation and new examples.
	
	Bug fixes
	* The 'acct_unique' module has been updated to understand
	  the deprecated (but still used) Client-IP-Address attribute.
	* The EAP-MSCHAPv2 module no longer leaks MS-CHAP2-Success in
	  packets.
	* Fixed crash in rlm_dbm.
	* Fixed parsing of syslog configuration.

FreeRADIUS 2.0.0 ; Date: 2007/11/24 08:33:09 , urgency=low
	Feature improvements
	* Debugging mode is much clearer and easier to read.
        * A new policy language makes many configurations trivial.
	  See "man unlang" for a complete description.
	* Virtual servers are now supported.  This permits clear separation
	  of policies.  See raddb/sites-available/README
	* EAP-TLS (PEAP, EAP-TTLS) and OpenSSL certificates "just work".
	  See raddb/certs/README for details.
	* Proxying is much more configurable than before.
	  See proxy.conf for documentation on pools, and new config items.
	* Full support for IPv6.
	* Much more complete support for the RADIUS SNMP MIBs.
	* HUP now works.  Only some modules are re-loaded,
	  and the server configuation is *not* reloaded.
	* "check config" option now works.  See "man radiusd"
	* radrelay functionality is now included in the server core.
	  See raddb/sites-available/copy-acct-to-home-server
	* VMPS support.  It is minimal, but functional.
	* Cleaned up internal API's and names, including library names.
	
	Bug fixes
	* Many.


src/modules/rlm_sql_log/rlm_sql_log.c
/*
 *  rlm_sql_log.c	Append the SQL queries in a log file which
 *			is read later by the radsqlrelay program
 *
 *  Version:    $Id$
 *
 *  Author:     Nicolas Baradakis <nicolas.baradakis@cegetel.net>
 *
 *  Copyright (C) 2005 Cegetel
 *  Copyright 2006 The FreeRADIUS server project
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA
 */

#include <freeradius-devel/ident.h>
RCSID("$Id$")

#include <freeradius-devel/radiusd.h>
#include <freeradius-devel/modules.h>

#include <fcntl.h>
#include <sys/stat.h>

static int sql_log_instantiate(CONF_SECTION *conf, void **instance);
static int sql_log_detach(void *instance);
static int sql_log_accounting(void *instance, REQUEST *request);
static int sql_log_postauth(void *instance, REQUEST *request);

#ifndef HAVE_PTHREAD_H
/*
 *      This is a lot simpler than putting ifdef's around
 *      every use of the pthread functions.
 */
#define pthread_mutex_lock(a)
#define pthread_mutex_trylock(a) (0)
#define pthread_mutex_unlock(a)
#define pthread_mutex_init(a,b)
#define pthread_mutex_destroy(a)
#else
#include	<pthread.h>
#endif

#define MAX_QUERY_LEN 4096

/*
 *	Define a structure for our module configuration.
 */
typedef struct rlm_sql_log_t {
	char		*name;
	char		*path;
	char		*postauth_query;
	char		*sql_user_name;
	char		*allowed_chars;
	CONF_SECTION	*conf_section;
#ifdef HAVE_PTHREAD_H
	pthread_mutex_t	mutex;
#endif
} rlm_sql_log_t;

/*
 *	A mapping of configuration file names to internal variables.
 */
static const CONF_PARSER module_config[] = {
	{"path", PW_TYPE_STRING_PTR,
	 offsetof(rlm_sql_log_t,path), NULL, "${radacctdir}/sql-relay"},
	{"Post-Auth", PW_TYPE_STRING_PTR,
	 offsetof(rlm_sql_log_t,postauth_query), NULL, ""},
	{"sql_user_name", PW_TYPE_STRING_PTR,
	 offsetof(rlm_sql_log_t,sql_user_name), NULL, ""},
	{"safe-characters", PW_TYPE_STRING_PTR,
	 offsetof(rlm_sql_log_t,allowed_chars), NULL,
	"@abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789.-_: /"},

	{ NULL, -1, 0, NULL, NULL }	/* end the list */
};

static char *allowed_chars = NULL;

/*
 *	Do any per-module initialization that is separate to each
 *	configured instance of the module.  e.g. set up connections
 *	to external databases, read configuration files, set up
 *	dictionary entries, etc.
 *
 *	If configuration information is given in the config section
 *	that must be referenced in later calls, store a handle to it
 *	in *instance otherwise put a null pointer there.
 */
static int sql_log_instantiate(CONF_SECTION *conf, void **instance)
{
	const char	*name;
	rlm_sql_log_t	*inst;

        /*
         *      Set up a storage area for instance data.
         */
        inst = calloc(1, sizeof(rlm_sql_log_t));
        if (inst == NULL) {
                radlog(L_ERR, "rlm_sql_log: Not enough memory");
                return -1;
        }

	/*
	 *      Get the name of the current section in the conf file.
	 */
	name = cf_section_name2(conf);
	if (name == NULL)
		name = cf_section_name1(conf);
	if (name == NULL)
		name = "sql_log";
	inst->name = strdup(name);

	/*
	 *	If the configuration parameters can't be parsed,
	 *	then fail.
	 */
	if (cf_section_parse(conf, inst, module_config) < 0) {
		radlog(L_ERR, "rlm_sql_log (%s): Unable to parse parameters",
		       inst->name);
		sql_log_detach(inst);
		return -1;
	}

	inst->conf_section = conf;
	allowed_chars = inst->allowed_chars;
	*instance = inst;
	return 0;
}

/*
 *	Say goodbye to the cruel world.
 */
static int sql_log_detach(void *instance)
{
	int i;
	char **p;
	rlm_sql_log_t *inst = (rlm_sql_log_t *)instance;

	if (inst->name) {
		free(inst->name);
		inst->name = NULL;
	}

	/*
	 *	Free up dynamically allocated string pointers.
	 */
	for (i = 0; module_config[i].name != NULL; i++) {
		if (module_config[i].type != PW_TYPE_STRING_PTR) {
			continue;
		}

		/*
		 *	Treat 'config' as an opaque array of bytes,
		 *	and take the offset into it.  There's a
		 *      (char*) pointer at that offset, and we want
		 *	to point to it.
		 */
		p = (char **) (((char *)inst) + module_config[i].offset);
		if (!*p) { /* nothing allocated */
			continue;
		}
		free(*p);
		*p = NULL;
	}
	allowed_chars = NULL;
	free(inst);
	return 0;
}

/*
 *	Translate the SQL queries.
 */
static size_t sql_escape_func(char *out, size_t outlen, const char *in)
{
	int len = 0;

	while (in[0]) {
		/*
		 *	Non-printable characters get replaced with their
		 *	mime-encoded equivalents.
		 */
		if ((in[0] < 32) ||
		    strchr(allowed_chars, *in) == NULL) {
			/*
			 *	Only 3 or less bytes available.
			 */
			if (outlen <= 3) {
				break;
			}

			snprintf(out, outlen, "=%02X", (unsigned char) in[0]);
			in++;
			out += 3;
			outlen -= 3;
			len += 3;
			continue;
		}

		/*
		 *	Only one byte left.
		 */
		if (outlen <= 1) {
			break;
		}

		/*
		 *	Allowed character.
		 */
		*out = *in;
		out++;
		in++;
		outlen--;
		len++;
	}
	*out = '\0';
	return len;
}

/*
 *	Add the 'SQL-User-Name' attribute to the packet.
 */
static int sql_set_user(rlm_sql_log_t *inst, REQUEST *request, char *sqlusername, const char *username)
{
	VALUE_PAIR *vp=NULL;
	char tmpuser[MAX_STRING_LEN];

	tmpuser[0] = '\0';
	sqlusername[0] = '\0';

	/* Remove any user attr we added previously */
	pairdelete(&request->packet->vps, PW_SQL_USER_NAME);

	if (username != NULL) {
		strlcpy(tmpuser, username, MAX_STRING_LEN);
	} else if (inst->sql_user_name[0] != '\0') {
		radius_xlat(tmpuser, sizeof(tmpuser), inst->sql_user_name,
			    request, NULL);
	} else {
		return 0;
	}

	if (tmpuser[0] != '\0') {
		strlcpy(sqlusername, tmpuser, sizeof(tmpuser));
		DEBUG2("rlm_sql_log (%s): sql_set_user escaped user --> '%s'",
		       inst->name, sqlusername);
		vp = pairmake("SQL-User-Name", sqlusername, 0);
		if (vp == NULL) {
			radlog(L_ERR, "%s", librad_errstr);
			return -1;
		}

		pairadd(&request->packet->vps, vp);
		return 0;
	}
	return -1;
}

/*
 *	Replace %<whatever> in the query.
 */
static int sql_xlat_query(rlm_sql_log_t *inst, REQUEST *request, const char *query, char *xlat_query, size_t len)
{
	char	sqlusername[MAX_STRING_LEN];

	/* If query is not defined, we stop here */
	if (query[0] == '\0')
		return RLM_MODULE_NOOP;

	/* Add attribute 'SQL-User-Name' */
	if (sql_set_user(inst, request, sqlusername, NULL) <0) {
		radlog(L_ERR, "rlm_sql_log (%s): Couldn't add SQL-User-Name attribute",
			       inst->name);
		return RLM_MODULE_FAIL;
	}

	/* Expand variables in the query */
	xlat_query[0] = '\0';
	radius_xlat(xlat_query, len, query, request, sql_escape_func);
	if (xlat_query[0] == '\0') {
		radlog(L_ERR, "rlm_sql_log (%s): Couldn't xlat the query %s",
		       inst->name, query);
		return RLM_MODULE_FAIL;
	}

	return RLM_MODULE_OK;
}

/*
 *	The Perl version of radsqlrelay uses fcntl locks.
 */
static int setlock(int fd)
{
	struct flock fl;
	memset(&fl, 0, sizeof(fl));
	fl.l_start = 0;
	fl.l_len = 0;
	fl.l_type = F_WRLCK;
	fl.l_whence = SEEK_SET;
	return fcntl(fd, F_SETLKW, &fl);
}

/*
 *	Write the line into file (with lock)
 */
static int sql_log_write(rlm_sql_log_t *inst, REQUEST *request, const char *line)
{
	int fd;
	FILE *fp;
	int locked = 0;
	struct stat st;
	char path[MAX_STRING_LEN];

	path[0] = '\0';
	radius_xlat(path, sizeof(path), inst->path, request, NULL);
	if (path[0] == '\0') {
		return RLM_MODULE_FAIL;
	}

	while (!locked) {
		if ((fd = open(path, O_WRONLY | O_APPEND | O_CREAT, 0666)) < 0) {
			radlog(L_ERR, "rlm_sql_log (%s): Couldn't open file %s: %s",
			       inst->name, path, strerror(errno));
			return RLM_MODULE_FAIL;
		}
		if (setlock(fd) != 0) {
			radlog(L_ERR, "rlm_sql_log (%s): Couldn't lock file %s: %s",
			       inst->name, path, strerror(errno));
			close(fd);
			return RLM_MODULE_FAIL;
		}
		if (fstat(fd, &st) != 0) {
			radlog(L_ERR, "rlm_sql_log (%s): Couldn't stat file %s: %s",
			       inst->name, path, strerror(errno));
			close(fd);
			return RLM_MODULE_FAIL;
		}
		if (st.st_nlink == 0) {
			DEBUG("rlm_sql_log (%s): File %s removed by another program, retrying",
			      inst->name, path);
			close(fd);
			continue;
		}
		locked = 1;
	}

	if ((fp = fdopen(fd, "a")) == NULL) {
		radlog(L_ERR, "rlm_sql_log (%s): Couldn't associate a stream with file %s: %s",
		       inst->name, path, strerror(errno));
		close(fd);
		return RLM_MODULE_FAIL;
	}
	fputs(line, fp);
	putc('\n', fp);
	fclose(fp);	/* and unlock */
	return RLM_MODULE_OK;
}

/*
 *	Write accounting information to this module's database.
 */
static int sql_log_accounting(void *instance, REQUEST *request)
{
	int		ret;
	char		querystr[MAX_QUERY_LEN];
	const char	*cfquery;
	rlm_sql_log_t	*inst = (rlm_sql_log_t *)instance;
	VALUE_PAIR	*pair;
	DICT_VALUE	*dval;
	CONF_PAIR	*cp;

	DEBUG("rlm_sql_log (%s): Processing sql_log_accounting", inst->name);

	/* Find the Acct Status Type. */
	if ((pair = pairfind(request->packet->vps, PW_ACCT_STATUS_TYPE)) == NULL) {
		radlog(L_ERR, "rlm_sql_log (%s): Packet has no account status type",
		       inst->name);
		return RLM_MODULE_INVALID;
	}

	/* Search the query in conf section of the module */
	if ((dval = dict_valbyattr(PW_ACCT_STATUS_TYPE, pair->vp_integer)) == NULL) {
		radlog(L_ERR, "rlm_sql_log (%s): Unsupported Acct-Status-Type = %d",
		       inst->name, pair->vp_integer);
		return RLM_MODULE_NOOP;
	}
	if ((cp = cf_pair_find(inst->conf_section, dval->name)) == NULL) {
		DEBUG("rlm_sql_log (%s): Couldn't find an entry %s in the config section",
		      inst->name, dval->name);
		return RLM_MODULE_NOOP;
	}
	cfquery = cf_pair_value(cp);

	/* Xlat the query */
	ret = sql_xlat_query(inst, request, cfquery, querystr, sizeof(querystr));
	if (ret != RLM_MODULE_OK)
		return ret;

	/* Write query into sql-relay file */
	return sql_log_write(inst, request, querystr);
}

/*
 *	Write post-auth information to this module's database.
 */
static int sql_log_postauth(void *instance, REQUEST *request)
{
	int		ret;
	char		querystr[MAX_QUERY_LEN];
	rlm_sql_log_t	*inst = (rlm_sql_log_t *)instance;

	DEBUG("rlm_sql_log (%s): Processing sql_log_postauth", inst->name);

	/* Xlat the query */
	ret = sql_xlat_query(inst, request, inst->postauth_query,
			     querystr, sizeof(querystr));
	if (ret != RLM_MODULE_OK)
		return ret;

	/* Write query into sql-relay file */
	return sql_log_write(inst, request, querystr);
}

/*
 *	The module name should be the only globally exported symbol.
 *	That is, everything else should be 'static'.
 *
 *	If the module needs to temporarily modify it's instantiation
 *	data, the type should be changed to RLM_TYPE_THREAD_UNSAFE.
 *	The server will then take care of ensuring that the module
 *	is single-threaded.
 */
module_t rlm_sql_log = {
	RLM_MODULE_INIT,
	"sql_log",
	RLM_TYPE_THREAD_UNSAFE | RLM_TYPE_CHECK_CONFIG_SAFE | RLM_TYPE_HUP_SAFE,		/* type */
	sql_log_instantiate,		/* instantiation */
	sql_log_detach,			/* detach */
	{
		NULL,			/* authentication */
		NULL,			/* authorization */
		NULL,			/* preaccounting */
		sql_log_accounting,	/* accounting */
		NULL,			/* checksimul */
		NULL,			/* pre-proxy */
		NULL,			/* post-proxy */
		sql_log_postauth	/* post-auth */
	},
};
